src/github.com/getlantern/flashlight/geolookup/geolookup.go

```diff
@@ -1,147 +1,103 @@
 package geolookup
 
 import (
-	"fmt"
 	"math"
-	"math/rand"
-	"sync"
-	"sync/atomic"
 	"time"
 
+	"github.com/getlantern/eventual"
 	geo "github.com/getlantern/geolookup"
 	"github.com/getlantern/golog"
 
-	"github.com/getlantern/flashlight/pubsub"
-	"github.com/getlantern/flashlight/ui"
 	"github.com/getlantern/flashlight/util"
 )
 
-const (
-	messageType = `GeoLookup`
-
-	basePublishSeconds     = 30
-	publishSecondsVariance = basePublishSeconds - 10
-	retryWaitMillis        = 100
-)
-
 var (
 	log = golog.LoggerFor("flashlight.geolookup")
 
-	service  *ui.Service
-	cfgMutex sync.Mutex
-	country  = atomicString()
-	ip       = atomicString()
-	cf       = util.NewChainedAndFronted()
-)
+	refreshRequest = make(chan eventual.Getter, 1)
+	currentGeoInfo = eventual.NewValue()
+	cf             util.HTTPFetcher
 
-func atomicString() atomic.Value {
-	var val atomic.Value
-	val.Store("")
-	return val
-}
+	retryWaitMillis = 100
+	maxRetryWait    = 30 * time.Second
+)
 
-func GetIp() string {
-	return ip.Load().(string)
+type geoInfo struct {
+	ip   string
+	city *geo.City
 }
 
-func GetCountry() string {
-	return country.Load().(string)
+// GetIP gets the IP. If the IP hasn't been determined yet, waits up to the
+// given timeout for an IP to become available.
+func GetIP(timeout time.Duration) string {
+	gi, ok := currentGeoInfo.Get(timeout)
+	if !ok || gi == nil {
+		return ""
+	}
+	return gi.(*geoInfo).ip
 }
 
-// Configure configures geolookup to use the given http.Client to perform
-// lookups. geolookup runs in a continuous loop, periodically updating its
-// location and publishing updates to any connected clients. We do this
-// continually in order to detect when the computer's location has changed.
-func Start() {
-	// Avoid annoying checks for nil later.
-	ip.Store("")
-	country.Store("")
-
-	if service == nil {
-		err := registerService()
-		if err != nil {
-			log.Errorf("Unable to register service: %s", err)
-			return
-		}
-		go write()
-		go read()
-		log.Debug("Running")
+// GetCountry gets the country. If the country hasn't been determined yet, waits
+// up to the given timeout for a country to become available.
+func GetCountry(timeout time.Duration) string {
+	gi, ok := currentGeoInfo.Get(timeout)
+	if !ok || gi == nil {
+		return ""
 	}
+	return gi.(*geoInfo).city.Country.IsoCode
 }
 
-func registerService() error {
-	helloFn := func(write func(interface{}) error) error {
-		country := GetCountry()
-		if country == "" {
-			log.Trace("No lastKnownCountry, not sending anything to client")
-			return nil
-		}
-		log.Trace("Sending last known location to new client")
-		return write(country)
+// Refresh refreshes the geolookup information by calling the remote geolookup
+// service. It will keep calling the service until it's able to determine an IP
+// and country.
+func Refresh(proxyAddrFN eventual.Getter) {
+	select {
+	case refreshRequest <- proxyAddrFN:
```

Comment:
So I believe this will re-queue a request as long as there's not a current one in-flight correct? In practice we only call Refresh once on startup AFAICT, and I'm not sure the existing behavior is necessarily wrong but maybe just not super intuitive (i.e. what are we really trying to do here? call this once? call it if an existing call isn't in flight?)